### DISCUSSION TITLE
How to combine different Data source  (API and PostgreSQL)

### DISCUSSION THREAD

**Discussion by Winniepoom** [Budibase/budibase#9723] [Help]
Hello I'm very new here and know little about coding. I hoped to get some help here.
So my main goal is to combine 2 different data sources, one is from Rest API (Employee data) and another one from PostgreSQL(HR Employee Data) . Is there any chance to combine it together or any workaround method.

Oh and one more question: if I manage to combine 2 data sources together, how do I bind some data like Employee name from 2 data sources to let know if name is matched they are the same person.

Thank you.

**Comment by melohagan** [comment]
Hey @Winniepoom 

The way to combine datasources would be through *soft links* between them in the front-end **Design** section using [Data providers](https://docs.budibase.com/docs/data-provider). 

For example, let's say your REST API returns employees with an **id** field. In addition the PostgreSQL source returns data with a key that could be used to join, e.g. **emp_id**.

Then you would have the following structure:
> Data provider (PostgreSQL - HR Employee Data)
> > Data provider (REST API - Employee Data)
> > > Repeater (loop over each employee from the REST API)
> > > > Card (or other display components)

The Card in this case could directly use bindings from the REST API response that is being looped over, but it could also access data from the PostgreSQL query with a little bit of JavaScript. Something like:
```js
return $("Postgres Data Provider Rows").find(row => row.emp_id == $("Repeater.REST API.id")).myPostgresColumnName
```

We do have plans to make this a little easier: https://github.com/Budibase/budibase/issues/9531

**Reply by andriykrefer** [reply]
@melohagan Could you help me understand what is being done under the hood?

- All postgres rows all fetched, and the filtering happens in the frontend. In this case, can large tables be an issue with data pagination limiting the query results?

or

- Each find() call triggers a new server-side query, so pagination is not an issue.

**Reply by melohagan** [reply]
Hey @andriykrefer In this particular case the JS find function would be filtering on the client-side. 

The reason that was done here is because the OP wanted to combine the results of two datasource (REST and Postgres), and this was an approach to do that.

With that said, this could also be done server side through automations: https://docs.budibase.com/docs/js-scripting#tutorial-linking-an-api-response-to-table-rows
